feat: add simple Copc reader

einwolkenfarmer opened the pull request:
Hey,

this pr adds copc headers.
There are some things to consider before proceeding with a full implementation:
- bounds and resolution logic
- ranged based requests
- laz reader needs more logic

Thanks,
Jonas

gadomski commented on Cargo.toml:
```diff
@@ -27,6 +27,7 @@ criterion = "0.5"
 [features]
 laz = ["dep:laz"]
 laz-parallel = ["dep:laz", "laz/parallel"]
+copc = ["laz"]
```
Since there's no extra dependencies, I don't think we need to add a new feature. We can just include the `copc` module if `laz` is enabled.

```suggestion
laz = ["dep:laz"]
laz-parallel = ["dep:laz", "laz/parallel"]
```

gadomski commented on src/copc.rs:
```diff
@@ -0,0 +1,395 @@
+    pub fn child(&self, dir: i32) -> Self {
+        // TODO: Maybe dir %= 8; would be better
+        if dir > 7 {
+            panic!("direction needs to be in range 0..8! Was {dir}")
```
Shouldn't panic, return an `Err` instead.

einwolkenfarmer replied:
Returns an error now. We might as well clamp or mod to the intended range instead?

gadomski replied:
🤔 will this be used on read? Or only when constructing a `Vlr`? If it's on read, then we should be permissive (so probably clamp and warn)? If it's used during constructing, and error would be better.

For background, when I initially wrote this crate I did reads exactly to the spec, which folks found annoying when reading slightly-incorrect files. I've since relaxed a lot of reading to accept invalid inputs: https://github.com/gadomski/las-rs/issues/63

einwolkenfarmer replied:
This will not be used on read. On write it would be used to compute Index nodes. 
If direction is outside of 0..8 the fault is with the user of this crate. It simply should not happen. We could abstract an enum for the direction to make sure of this, but I hate naming enum variants x0y0z0 to x1y1z1 :D
Maybe I'm overthinking this and it should just be clamped.

I agree with you on the permissive read. Especially with Copc-writers the output files seem to diverge some. So we should be permissive here too.

gadomski replied:
> If direction is outside of 0..8 the fault is with the user of this crate.

Let's return an `Err` for now, and if folks complain we can always relax it later.

gadomski commented on src/copc.rs:
```diff
@@ -0,0 +1,395 @@
+impl VoxelKey {
+    /// Computes the Childs of a VoxelKey
+    /// There are max 8 Childs to a VoxelKey
+    /// **dir**ection needs to be 0..8
+    pub fn child(&self, dir: i32) -> Self {
```
Letters are cheap, let's avoid abbreviations if we can.

```suggestion
    /// **dir**ection needs to be 0..8
    pub fn child(&self, direction: i32) -> Self {
```

gadomski commented on src/copc.rs:
```diff
@@ -0,0 +1,395 @@
+//! copc header data
+//!
```
```suggestion
//! [COPC](https://copc.io/) header data
```

gadomski commented on src/copc.rs:
```diff
@@ -0,0 +1,395 @@
+    }
+}
+
+/// An entry corresponds to a single key/value pair in an EPT hierarchy, but con-
+/// tains additional information to allow direct access and decoding of the corre-
```
Shouldn't have hyphenated line breaks, let's wrap on whitespace.

gadomski commented on src/copc.rs:
```diff
@@ -0,0 +1,395 @@
+}
+impl CopcHierarchyVlr {
+    /// The user id of the LasZip VLR header.
+    pub const USER_ID: &'static str = "copc";
+    /// The record id of the LasZip VLR header.
+    pub const RECORD_ID: u16 = 1000;
+    /// The description of the LasZip VLR header.
+    pub const DESCRIPTION: &'static str = "https://copc.io";
```
The user id and description are duplicated between the two VLR types, so let's make them module-level constants.

gadomski commented on src/copc.rs:
````diff
@@ -0,0 +1,395 @@
+/// assert!(copc::is_copcinfo_vlr(&vlr));
+/// }
+/// ```
+pub fn is_copcinfo_vlr(vlr: &Vlr) -> bool {
+    vlr.user_id == CopcInfoVlr::USER_ID && vlr.record_id == CopcInfoVlr::RECORD_ID
+}
````
```suggestion
impl Vlr {
pub fn is_copc_info(&self) -> bool {
    self.user_id == CopcInfoVlr::USER_ID && self.record_id == CopcInfoVlr::RECORD_ID
   }
}
```

gadomski commented on src/copc.rs:
```diff
@@ -0,0 +1,395 @@
+
+impl Header {
+    /// doc
+    pub fn copcinfo_vlr(&self) -> Result<CopcInfoVlr> {
```
```suggestion
    pub fn copc_info_vlr(&self) -> Result<CopcInfoVlr> {
```

gadomski commented on src/copc.rs:
```diff
@@ -0,0 +1,395 @@
+    }
+}
+
+/// Returns true if this [Vlr] is the Copc Page Vlr.
```
```suggestion
/// Returns true if this [Vlr] is the Copc Heirarchy Vlr.
```

gadomski commented on src/copc.rs:
````diff
@@ -0,0 +1,395 @@
+/// assert!(copc::is_copchierarchy_evlr(&vlr));
+/// }
+/// ```
+pub fn is_copchierarchy_evlr(vlr: &Vlr) -> bool {
+    vlr.user_id == CopcHierarchyVlr::USER_ID && vlr.record_id == CopcHierarchyVlr::RECORD_ID
+}
````
Same as above, let's make this a method on `Vlr`

gadomski commented on src/copc.rs:
```diff
@@ -0,0 +1,395 @@
+
+impl Header {
+    /// doc
+    pub fn copchierarchy_evlr(&self) -> Result<CopcHierarchyVlr> {
```
```suggestion
    pub fn copc_hierarchy_evlr(&self) -> Result<CopcHierarchyVlr> {
```

gadomski reviewed (changes requested):
Thanks for the contribution! Looks like there's some formatting nits, you can fix those locally w/ `cargo +nightly fmt`.

Requested a few refactoring-style changes but generally feels about right.

einwolkenfarmer commented:
Thanks for the nice review! Just one small question left.
I think I'll try to implement the reader with VoxelKeys only. The expectation should be that the User writes his own LOD logic. If we try it here we would need to pull an octree, raycast and perspective view dependency to queue the correct Voxels. And even then the user might want a simpler implementation.

gadomski commented:
👍🏼 to keeping things simple for now and letting downstreams decide how they want to use the header.

gadomski commented on src/copc.rs:
```diff
@@ -0,0 +1,428 @@
+            .unwrap();
+        assert_eq!(root_entry.key, VoxelKey::ROOT);
+        assert_eq!(root_entry.point_count, 107);
+        println!("{root_entry:?}");
```
```suggestion
```

gadomski commented on src/copc.rs:
```diff
@@ -0,0 +1,427 @@
+    /// **dir**ection needs to be 0..8
+    pub fn child(&self, direction: i32) -> Result<Self> {
+        // TODO: Maybe direction %= 8; would be better
+        if !(0..8).contains(&direction) {
```
```suggestion
        if direction >= 0 && direction <= 8 {
```

einwolkenfarmer replied:
clippy complains about a manual range contains implementation 😄

gadomski replied:
Right on! Clippy is usually right 😊

einwolkenfarmer commented:
Hey, I added a reader for a Copc Entry. There is no good way to use the ReadPoint trait for this so i refactored the header reading and building to construct a CopcEntryReader from a raw stream.

Closes #57

gadomski commented on src/copc.rs:
```diff
@@ -0,0 +1,427 @@
+    pub fn child(&self, direction: i32) -> Result<Self> {
+        // TODO: Maybe direction %= 8; would be better
+        if !(0..8).contains(&direction) {
+            return Err(Error::FunctionArgumentRequirementsNotMet {
+                argument: format!("direction needs to be (0..8). Was {direction}"),
+            });
```
This feels like too-general of an error ... I think I'd prefer `Error::InvalidDirection(direction)`.

gadomski commented on src/header/mod.rs:
```diff
@@ -106,6 +112,86 @@ pub struct Header {
 
 impl Header {
+    /// Reads all header, vlr and evlr data from file and returns the complete header
+    pub fn read_and_build_from<R: Read + Seek>(mut read: R) -> Result<Self> {
```
In this repo, we've used `new` to build structure from a `Read`:

```suggestion
    pub fn new<R: Read + Seek>(mut read: R) -> Result<Self> {
```

gadomski commented on src/reader/copc.rs:
```diff
@@ -0,0 +1,90 @@
+    }
+}
+#[cfg(test)]
+mod test {
```
```suggestion
mod tests {
```

gadomski commented on src/reader/copc.rs:
I _think_ I'd prefer to keep all the `copc` stuff in a single module, the `src/copc` file?

einwolkenfarmer commented:
Should be fine now, but I'll let it rest for a bit and check in again after the holidays. 🐰 
You can merge if you think it's fine, though. 
Happy Easter!

gadomski commented on src/copc.rs:
```diff
@@ -0,0 +1,507 @@
+                if let Some(page) = self.sub_pages.get(&entry.key) {
+                    page.entries.clone()
+                } else {
+                    // this entry is corrupt or the page is missing
+                    vec![Entry {
+                        key: entry.key,
+                        offset: entry.offset,
+                        byte_size: 0,
+                        point_count: 0,
+                    }]
```
Do we want to iterate over `Result<Entry>` and return an error in this case?

gadomski commented on src/copc.rs:
```diff
@@ -0,0 +1,507 @@
+        self.root.entries.iter().flat_map(|entry| {
+            if entry.is_referencing_page() {
+                if let Some(page) = self.sub_pages.get(&entry.key) {
+                    page.entries.clone()
```
Should we iterate over references so we don't need to clone?

gadomski commented on src/copc.rs:
```diff
@@ -0,0 +1,507 @@
+        Ok(CopcHierarchyVlr { root, sub_pages })
+    }
+    /// iterates over all entries merging all referenced pages into root
+    pub fn iter_entrys(&self) -> impl Iterator<Item = Entry> {
```
```suggestion
    pub fn iter_entries(&self) -> impl Iterator<Item = Entry> {
```

gadomski commented on src/copc.rs:
```diff
@@ -0,0 +1,507 @@
+}
+
+impl Header {
+    /// doc
```
Needs some docs

gadomski commented on src/copc.rs:
```diff
@@ -0,0 +1,507 @@
+
+impl Header {
+    /// doc
+    pub fn copc_info_vlr(&self) -> Result<CopcInfoVlr> {
```
I _think_ I'm more inclined to return an option, and let the caller decide how bad that is:

```suggestion
    pub fn copc_info_vlr(&self) -> Option<CopcInfoVlr> {
```

gadomski commented on src/copc.rs:
```diff
@@ -0,0 +1,507 @@
+    }
+}
+impl Header {
+    /// doc
```
Needs doc, and same as above I think we should just return an `Option`.

gadomski reviewed (changes requested):
A few more design notes, but feels really good, think it's close. Really appreciate the iterations!

einwolkenfarmer commented:
Great reviews, thank you very much! Refactoring the EntryReader was a bit complicated but I think it turned out well.
I had some help generating the docs 😉  but you can probably tell.

gadomski reviewed (approved):
Looks great! I did some light docs cleanup, let's :shipit:. Really appreciate the contribution, this is awesome!
